### PR TITLE
e2e: schedule IBC software upgrade

### PR DESCRIPTION
**Pull request by charleenfei**
## Description

closes: #XXXX

### Commit Message / Changelog Entry

```bash
type: commit message
```

see the [guidelines](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#commit-messages) for commit messages. (view raw markdown for examples)

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [ ] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#pull-request-targeting)).
- [ ] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/main/docs/docs/build/building-modules/11-structure.md) and [Go style guide](../docs/dev/go-style-guide.md).
- [ ] Wrote unit and integration [tests](https://github.com/cosmos/ibc-go/blob/main/testing/README.md#ibc-testing-package).
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`).
- [ ] Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Provide a [commit message](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#commit-messages) to be used for the changelog entry in the PR description for review.
- [ ] Re-reviewed `Files changed` in the Github PR explorer.
- [ ] Review `Codecov Report` in the comment section below once CI passes.

### CODE REVIEW

File: e2e/tests/core/02-client/client_test.go
Comment (charleenfei) [inline e2e/tests/core/02-client/client_test.go:202]: I wanted to query if the plan has been stored/upgraded client state was stored, but realised after digging around that the UpgradeKeeper doesn't seem to have a query client. what do you guys think? our `msg_server` tests do perform this check, is that sufficient?

Comment (charleenfei) [reply e2e/tests/core/02-client/client_test.go:202]: another option would be to wait for the upgrade plan to pass (`time.Sleep` the requisite amount of time, then check the trusting periods).

File: e2e/tests/core/02-client/client_test.go
Comment (colin-axner) [inline e2e/tests/core/02-client/client_test.go:166]: what's this for?

File: e2e/tests/core/02-client/client_test.go
Comment (colin-axner) [inline e2e/tests/core/02-client/client_test.go:180]: the authority is already obtained on line 168. This can be removed?

File: e2e/tests/core/02-client/client_test.go
Comment (colin-axner) [reply e2e/tests/core/02-client/client_test.go:202]: You can try to upgrade the client on the counterparty. Hermes might do it automatically. Maybe try starting the relayer and querying the client's status afterward

File: e2e/tests/core/02-client/client_test.go
Comment (colin-axner) [inline e2e/tests/core/02-client/client_test.go:182]: you should query the client on chainB if you upgrade chainA

File: e2e/tests/core/02-client/client_test.go
Comment (colin-axner) [inline e2e/tests/core/02-client/client_test.go:191]: might be nicer to reference the actual chain height, can we obtain it through chainA?

File: e2e/tests/core/02-client/client_test.go
Comment (charleenfei) [reply e2e/tests/core/02-client/client_test.go:166]: ah, an erroneous wait statement!

File: e2e/tests/core/02-client/client_test.go
Comment (damiannolan) [reply e2e/tests/core/02-client/client_test.go:197]: > You can try to upgrade the client on the counterparty. Hermes might do it automatically. Maybe try starting the relayer and querying the client's status afterward

Might need to check if interchaintest has built in the support for client updates. Hermes can do it based on the links in my first comment. But I'm unsure if rly also has the same support, and interchaintest would need to expose an API for it in their relayer interface.

Comment (colin-axner) [reply e2e/tests/core/02-client/client_test.go:197]: We added event emission when the consensus state is written allowing the counterparty client to be upgraded, unsure if hermes started listening for that event

File: e2e/tests/core/02-client/client_test.go
Comment (colin-axner) [inline e2e/tests/core/02-client/client_test.go:75]: this test can be deleted, likely added from merge conflicts

File: e2e/tests/core/02-client/client_test.go
Comment (colin-axner) [reply e2e/tests/core/02-client/client_test.go:115]: I regress on my suggestion, happy to finish the full e2e of upgrading the client in a followup pr as this isn't essential for v8 testing

Comment (charleenfei) [reply e2e/tests/core/02-client/client_test.go:115]: after discussion with colin, will just merge this e2e testing the scope of the `ScheduleIBCSoftwareUpgrade` for now.

File: e2e/tests/core/02-client/client_test.go
Comment (colin-axner) [inline e2e/tests/core/02-client/client_test.go:122]:
```suggestion
		s.Require().Equal(newTrustingPeriod, upgradedClientState.TrustingPeriod)
```

File: e2e/tests/core/02-client/client_test.go
Comment (colin-axner) [inline e2e/tests/core/02-client/client_test.go:128]: might want to assign to local var at top of the file

**Review by colin-axner — APPROVED** [review]
Thank you @charleenfei!!

File: e2e/tests/core/02-client/client_test.go
Comment (crodriguezvega) [reply e2e/tests/core/02-client/client_test.go:115]: > I think their cmd would need to change for gov v1, maybe it still uses the legacy v1beta1 appraoch

You're right, @damiannolan: hermes is using [gov v1beta proposal](https://github.com/informalsystems/hermes/blob/b7377772a3c9611cff25c5d8cf7ed9a5a26cf992/crates/relayer/src/upgrade_chain.rs#L71-L168). Let's ask them.

File: e2e/tests/core/02-client/client_test.go
Comment (colin-axner) [inline e2e/tests/core/02-client/client_test.go:87]: is this upgrading the chain to have an empty chainID? We should set some value?

Comment (damiannolan) [reply e2e/tests/core/02-client/client_test.go:87]: looks like it

**Review by damiannolan — APPROVED** [review]
LGTM pending chain ID using some value

File: e2e/tests/core/02-client/client_test.go
Comment (chatton) [inline e2e/tests/core/02-client/client_test.go:101]: seems we are doing a mix of `Require` and `Assert`. Maybe we can do `Require().NotEqual(...)` here to stay consistent?

File: e2e/tests/core/02-client/client_test.go
Comment (chatton) [inline e2e/tests/core/02-client/client_test.go:100]: maybe this could be a constant at the top of the test? I don't think there's a need to declare it at the top then assign here.

File: e2e/tests/core/02-client/client_test.go
Comment (chatton) [reply e2e/tests/core/02-client/client_test.go:87]: I think this should be a constant defined here

**Review by chatton — APPROVED** [review]
LGTM! Just a few small suggestions 👍 

Also - seems the linter is complaining 😢

File: e2e/tests/core/02-client/client_test.go
Comment (colin-axner) [reply e2e/tests/core/02-client/client_test.go:87]: I asked @charleenfei to update this to pull the current chainID and increment its revision number (as this is the recommended approach to changing chainIDs)

**Pull request merged**
